### PR TITLE
Check bytecodes against constants

### PR DESCRIPTION
**Pull request by nambrot**
### Description

This PR makes the checker check the on-chain bytecode against hardcoded hash of what is expected for a given contract and will log a violation if it does not match. This was chosen as a reasonable step forward without building large machinery to compare it against source code.

Runs clean against mainnet2 and testnet3.

### CODE REVIEW

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts [hunk 1]
```diff
@@ -15,6 +17,15 @@ import {
   ThresholdViolation,
 } from './types';
 
+const MAILBOX_WITHOUT_LOCAL_DOMAIN_BYTE_CODE_HASH =
```
Comment (asaj) [inline typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:20]: How were these generated?

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts [hunk 2]
```diff
@@ -63,6 +74,23 @@ export class HyperlaneCoreChecker<
       };
       this.addViolation(violation);
     }
+
+    await this.checkBytecodeHash(
```
Comment (asaj) [inline typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:78]: Consider moving all of these checks into a single function

**Review by asaj — COMMENTED** [review]
Thanks for doing this!

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts (hunk 1 above)
Comment (nambrot) [reply typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:20]: I ran the checker and took the `actual` values from the violations

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts [hunk 3]
```diff
@@ -63,6 +74,23 @@ export class HyperlaneCoreChecker<
...
+      chain,
+      'Mailbox implementation',
+      contracts.mailbox.addresses.implementation,
+      MAILBOX_WITHOUT_LOCAL_DOMAIN_BYTE_CODE_HASH,
+      (_) =>
+        // This is obviously super janky but basically we are searching
```
[the remark this answers is not in the document]
Comment (nambrot) [reply typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:84]: That's what I thought, but IIRC there were actually 3 occurrences of the localDomain (or 4 in the case of polygon)

File: typescript/sdk/src/deploy/HyperlaneAppChecker.ts [hunk 4]
```diff
@@ -77,6 +84,35 @@ export abstract class HyperlaneAppChecker<
     }
   }
 
+  private removeBytecodeMetadata(bytecode: string): string {
+    // https://docs.soliditylang.org/en/v0.8.17/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
+    // Remove solc metadata from bytecode
+    return bytecode.substring(0, bytecode.length - 90);
+  }
+
+  async checkBytecodeHash(
```
Comment (asaj) [inline typescript/sdk/src/deploy/HyperlaneAppChecker.ts:93]: Ditto, we are checking the bytecode, using the hash is just the mechanism

File: typescript/sdk/src/deploy/HyperlaneAppChecker.ts [hunk 5]
```diff
@@ -77,6 +84,35 @@ export abstract class HyperlaneAppChecker<
...
+    chain: Chain,
+    name: string,
+    address: string,
+    expectedByteCodeHash: string,
```
Comment (asaj) [inline typescript/sdk/src/deploy/HyperlaneAppChecker.ts:97]: nit: `ByteCode` vs `Bytecode`, be consistent (suggest `Bytecode`)

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts [hunk 6]
```diff
@@ -82,6 +169,7 @@ export class HyperlaneCoreChecker<
       contracts.interchainGasPaymaster.addresses,
       contracts.proxyAdmin.address,
     );
+
```
Comment (asaj) [inline typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:172]: nit: remove?

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts [hunk 7]
```diff
@@ -68,6 +83,78 @@ export class HyperlaneCoreChecker<
     }
   }
 
+  async checkBytecodeHashes(chain: Chain): Promise<void> {
```
Comment (asaj) [inline typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:86]: Do you have a sense of how these violations would be remediated? In other words, is the goal here just to flag these violations? Or to fix them?

**Review by asaj — APPROVED** [review]

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts (hunk 7 above)
Comment (asaj) [reply typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:86]: I guess I'm just not totally clear what the point of this is

Comment (nambrot) [reply typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:86]: Added a comment to that effect on the method

**Pull request merged**